contracts/Joint.sol

```diff
@@ -159,85 +185,302 @@ contract Joint {
+
+                if (sellToken == tokenA) {
+                    currentA = currentA.sub(sellAmount);
+                    currentB = currentB.add(buyAmount);
+                } else {
+                    currentB = currentB.sub(sellAmount);
+                    currentA = currentA.add(buyAmount);
+                }
+
+                (ratioA, ratioB) = getRatios(
+                    currentA,
+                    currentB,
+                    investedA,
+                    investedB
+                );
+                emit Ratios(ratioA, ratioB, "after balance");
+            }
+        }
 
-    function harvest() external onlyGuardians {
-        // IF tokenA or tokenB are rewards, we would be swapping all of it
-        // Let's save the previous balance before claiming
-        uint256 previousBalanceOfReward = balanceOfReward();
+        investedA = investedB = 0;
 
-        // Gets the reward from the masterchef contract
-        getReward();
-        uint256 rewardProfit = balanceOfReward().sub(previousBalanceOfReward);
-        if (rewardProfit > 0) {
-            swapReward(rewardProfit);
+        if (returnFunds) {
+            _returnLooseToProviders();
         }
+    }
 
+    function adjustPosition(bool reinvest) external onlyProviders {
         // No capital, nothing to do
-        if (balanceOfA() == 0 && balanceOfB() == 0) {
+        if (balanceOfA() == 0 || balanceOfB() == 0) {
             return;
         }
 
         if (reinvest) {
-            createLP();
+            require(
+                balanceOfStake() == 0 &&
+                    balanceOfPair() == 0 &&
+                    investedA == 0 &&
+                    investedB == 0
+            ); // don't create LP if we are already invested
+
+            (investedA, investedB, ) = createLP();
             depositLP();
+        }
+    }
+
+    function estimatedTotalAssetsAfterBalance()
+        public
+        view
+        returns (uint256 _aBalance, uint256 _bBalance)
+    {
+        uint256 rewardsPending = pendingReward();
+
+        (_aBalance, _bBalance) = balanceOfTokensInLP();
+
+        if (reward == tokenA) {
+            _aBalance = _aBalance.add(rewardsPending);
+        } else if (reward == tokenB) {
+            _bBalance = _bBalance.add(rewardsPending);
+        } else if (rewardsPending != 0) {
+            address swapTo = findSwapTo(reward);
+            uint256[] memory outAmounts =
+                IUniswapV2Router02(router).getAmountsOut(
+                    rewardsPending,
+                    getTokenOutPath(reward, swapTo)
+                );
+            if (swapTo == tokenA) {
+                _aBalance = _aBalance.add(outAmounts[outAmounts.length - 1]);
+            } else if (swapTo == tokenB) {
+                _bBalance = _bBalance.add(outAmounts[outAmounts.length - 1]);
+            }
+        }
+
+        (address sellToken, uint256 sellAmount) =
+            calculateSellToBalance(_aBalance, _bBalance, investedA, investedB);
```

Comment:
I might be reading it wrong but seems like you're incurring unnecessary swap fees (.3% of half the reward or .15% of total rewards) here by selling twice: `reward -> tokenA -> half A to B` when you could just do `reward -> tokenA & tokenB evenly`

Reply:
https://github.com/poolpitako/joint-strategy/pull/15#discussion_r689157035

Reply:
> yes. it is not that simple as splitting rewards in two as tokenA and tokenB are unbalanced at this point but it can be done more efficiently.
> 
> if this was a thing of saving gas I wouldn't care but this can actually increase APY a little bit (as you pay less swap fees) so it might be worth it to think a bit through it

current state being unbalanced vs splitting rewards seems like 2 separate things. 
I'm talking about the splitting. If there's `$100 Reward`, it should be sold to `$50 R -> $49.85 A  & $50 R -> $49.85 B`
instead of 
`$100 R -> $99.7 A`, then half of that `$49.85 A -> $49.70045 B`

Reply:
They reward splitting and PnL balancing are effectively the same thing. In the case of $100 of sushi rewards and LPing $100 worth of yfi and eth, we current: 

1. Swap $100 sushi for $100 of eth
2. We have $200 of eth and $100 of yfi, so we sell $50 eth for yfi to balance

The alternative case,
1. We swap $50 of sushi for eth
2. And $50 of sushi for yfi via eth (we pay fees on both steps sushi=>eth and eth=>yfi)

These two processes are equivalent fee wise.

Reply:
This only works if one side of the LP is weth

Reply:
> They reward splitting and PnL balancing are effectively the same thing. In the case of $100 of sushi rewards and LPing $100 worth of yfi and eth, we currently:
> 
> 1. Swap $100 sushi for $100 of eth
> 2. We have $200 of eth and $100 of yfi, so we sell $50 eth for yfi to balance
> 
> The alternative case,
> 
> 1. We swap $50 of sushi for eth
> 2. And $50 of sushi for yfi via eth (we pay fees on both steps sushi=>eth and eth=>yfi)


Using this example but imagine it's not yfi-eth lp. Say it's yfi-wbtc lp instead
1. Swap $100 sushi for $100 of _wbtc_. `sushi -> weth -> wbtc`
2. We have $200 of _wbtc_ and $100 of yfi, so we sell $50 _wbtc_ for yfi to balance `wbtc -> weth -> yfi`

Your volume traded is effectively $150 here

The alternative case,
We swap $50 of sushi for _wbtc_ `sushi -> weth -> wbtc`
And $50 of sushi for yfi via eth (we pay fees on both steps sushi=>eth and eth=>yfi) `sushi -> weth -> yfi`

Volume traded is only $100 here

> These two processes are equivalent fee wise. 
It's not in general cases. It only is equivalent for your example bc one of the lp pairs is weth which allowed you to skip a step in the trading path. More volume traded = more fees, alternative case should be more effective

Reply:
No disagreement that the current solution is not optimal if one side isn't weth or the reward token. For now, I'm only targeting pairs where one side is weth.